python/dgl/node_flow.py

```diff
@@ -0,0 +1,408 @@
+        raise DGLError('Readonly graph. Mutation is not allowed.')
+
+    def add_edge(self, u, v, data=None):
+        """Add one edge. Disabled because BatchedDGLGraph is read-only."""
+        raise DGLError('Readonly graph. Mutation is not allowed.')
+
+    def add_edges(self, u, v, data=None):
+        """Add many edges. Disabled because BatchedDGLGraph is read-only."""
+        raise DGLError('Readonly graph. Mutation is not allowed.')
+
+    def _get_layer_id(self, layer_id):
+        if layer_id >= 0:
+            return layer_id
+        else:
+            return self.num_layers + layer_id
+
+    def _get_flow_id(self, flow_id):
+        if flow_id >= 0:
+            return flow_id
+        else:
+            return self.num_flows + flow_id
+
+    def _get_node_frame(self, layer_id):
+        return self._node_frames[layer_id]
+
+    def _get_edge_frame(self, flow_id):
+        return self._edge_frames[flow_id]
+
+    @property
+    def num_layers(self):
+        """Get the number of layers.
+
+        Returns
+        -------
+        int
+            the number of layers
+        """
+        return len(self._layer_offsets) - 1
+
+    @property
+    def num_flows(self):
+        return self.num_layers - 1
+
+    @property
+    def layers(self):
+        return LayerView(self)
+
+    @property
+    def flows(self):
+        return FlowView(self)
+
+    def layer_size(self, layer_id):
+        """Return the number of nodes in a specified layer."""
+        layer_id = self._get_layer_id(layer_id)
+        return self._layer_offsets[layer_id + 1] - self._layer_offsets[layer_id]
+
+    def flow_size(self, flow_id):
+        flow_id = self._get_flow_id(flow_id)
+        return self._flow_offsets[flow_id + 1] - self._flow_offsets[flow_id]
+
+    def copy_from_parent(self):
+        """Copy node/edge features from the parent graph.
+
+        All old features will be removed.
+        """
+        # TODO we need to avoid copying the same node embedding to all layers.
+        if self._parent._node_frame.num_rows != 0:
+            for i in range(self.num_layers):
+                nid = utils.toindex(self.layer_parent_nid(i))
+                self._node_frames[i] = FrameRef(Frame(self._parent._node_frame[nid]))
+        if self._parent._edge_frame.num_rows != 0:
+            for i in range(self.num_flows):
+                eid = utils.toindex(self.flow_parent_eid(i))
+                self._edge_frames[i] = FrameRef(Frame(self._parent._edge_frame[eid]))
+
+    def copy_to_parent(self):
+        """Copy node/edge embeddings to the parent graph.
+        """
+        #TODO We need to take care of the following things:
+        #    * copy right node embeddings back. For instance, we should copy temporary
+        #      node embeddings back; we don't need to copy read-only node embeddings back.
+        #    * When nodes in different layers have the same node embedding, we need
+        #      to avoid conflicts.
```

Comment:
Do we really need to handle these complicated cases? For example, in which method will there be conflicts (when nodes in different layers have the same node embedding)?  In Control Variate method, each node has `num_layers` different history which need to be copy to parent.

I think it is better and more flexible to let users specify which keys to copy back to the parent graph.


Then, could we implement as

```
def copy_to_parent(self, layer_id_to_keys):
    for i in range(self.num_layers):
        nid = self.map_to_parent_nid(self.layer_parent_nid(i))
        self._parent._node_frame.update_rows(nid,
            {k :  self._node_frames[i][k] for k in layer_id_to_keys[i]}, inplace=True)
        
```

Reply:
I agree. I think it's easier just to prefix the feature name with `l0/`, `l1/`, .... The prefix can be configured by the user.